pkg/signature/algorithm_registry.go

```diff
@@ -0,0 +1,212 @@
+//
+// Copyright 2024 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package signature
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/elliptic"
+	"crypto/rsa"
+	"fmt"
+
+	v1 "github.com/sigstore/protobuf-specs/gen/pb-go/common/v1"
+)
+
+// PublicKeyType represents the public key algorithm for a given signature algorithm.
+type PublicKeyType uint
+
+const (
+	RSA     PublicKeyType = iota // RSA public key
+	ECDSA                        // ECDSA public key
+	ED25519                      // ED25519 public key
+)
+
+// RSAKeySize represents the size of an RSA public key.
+type RSAKeySize int
+
+// AlgorithmDetails exposes relevant information for a given signature algorithm.
+type AlgorithmDetails interface {
```

Comment:
I made this interface because I don't think it makes sense for users to be constructing their own.

Reply:
Do you think users will ever need to access this? I don't see functions that return an `AlgorithmDetails` (apart from `GetAlgorithmDetails` which I think can be made private).

I'd be more for keeping this private unless we see a real use case for having all these info. It should also be easier to add new public APIs in the future if we need them rather than modifying already public interfaces/structures which would require major versions updates.

Reply:
Of course if we do make it private we could just remove it altogether and keep only the `Impl` struct for now.

Reply:
Yes, I *think* we can make this private for now.

Reply:
I thought we agreed that we wanted to expose a way of getting the hash function, key type, etc for a given signature algorithm, unless I misunderstood that conversation.

Reply:
Thinking this a bit more and you are probably right @tetsuo-cpp , we might need this after all. At least, it sounds like a good type to move around between functions that need to know what kind of signing algorithm needs to be applied.

For example, in cosign there is this code:
```golang
func signerFromNewKey() (*SignerVerifier, error) {
	privKey, err := cosign.GeneratePrivateKey()
	if err != nil {
		return nil, fmt.Errorf("generating cert: %w", err)
	}
	sv, err := signature.LoadECDSASignerVerifier(privKey, crypto.SHA256)
	if err != nil {
		return nil, err
	}

	return &SignerVerifier{
		SignerVerifier: sv,
	}, nil
}
```

And it will likely need a new argument (coming down from the CLI flag) that specifies which type of private key to use (and its arguments) and the hash function to use with the Signer/Verifier.

So that could probably become:
```golang
func signerFromNewKey(algorithmDetails AlgorithmDetails) (*SignerVerifier, error) {
	privKey, err := cosign.GeneratePrivateKey(algorithmDetails)
	if err != nil {
		return nil, fmt.Errorf("generating cert: %w", err)
	}
	sv, err := signature.LoadSignerVerifier(privKey, algorithmDetails.GetHashType())
	if err != nil {
		return nil, err
	}

	return &SignerVerifier{
		SignerVerifier: sv,
	}, nil
}
```
and then the `GeneratePrivateKey` can do something like this:
```golang
	switch signingAlgorithm.GetKeyType() {
	case signature.ECDSA:
		curve, err := signingAlgorithm.GetECDSACurve()
		if err != nil {
			return nil, err
		}
		return ecdsa.GenerateKey(*curve, rand.Reader)
	case signature.ED25519:
		_, priv, err := ed25519.GenerateKey(rand.Reader)
		return priv, err
	default:
		return nil, fmt.Errorf("unsupported signing algorithm: %s", signingAlgorithm)
	}
```